tests/posix/common/src/pthread.c

```diff
@@ -992,3 +992,51 @@ ZTEST(posix_apis, test_pthread_attr_large_stacksize)
 	zassert_equal(actual_size, expect_size);
 	zassert_ok(pthread_attr_destroy(&attr));
 }
+
+static void *test_pthread_cancel_fn(void *arg)
+{
+	int oldtype = ~PTHREAD_CANCEL_DEFERRED;
+	int oldstate = ~PTHREAD_CANCEL_ENABLE;
+
+	zassert_ok(pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &oldtype));
+	zassert_equal(oldtype, PTHREAD_CANCEL_DEFERRED);
+
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &oldstate));
+	zassert_equal(oldstate, PTHREAD_CANCEL_ENABLE);
+
+	/* this should be ignored */
+	pthread_testcancel();
+
+	/* this will mark it pending */
+	pthread_cancel(pthread_self());
+
+	/* assume this completes well before 500ms */
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &oldstate));
+
+	/* this should terminate the thread*/
+	pthread_testcancel();
+
+	/*
+	 * We use this large delay to indicate that the thread in fact did not exit via
+	 * pthread_testcancel(), as expected. I.e. it should not take > 1000 ms to join thread
+	 */
+	zassert_ok(k_msleep(500));
+
+	return NULL;
+}
+
+ZTEST(posix_apis, test_pthread_testcancel)
+{
+	uint32_t now;
+	uint32_t then;
+	pthread_t th;
+
+	then = k_uptime_get_32();
+	zassert_ok(pthread_create(&th, NULL, test_pthread_cancel_fn, NULL));
+	zassert_ok(k_msleep(500));
+	zassert_ok(pthread_join(th, NULL));
+	now = k_uptime_get_32();
+
+	/* should not take > 1000 ms to join thread */
+	zassert_true(now - then <= 1000);
```

Comment:
if the main thread and the test thread runs in parallel, seems like we are going to meet this condition regardless of whether `pthread_testcancel` works properly?

Reply:
You may have right.. That why we need code review 😅

Reply:
Maybe I should check on the join part, if the thread has already terminate itself it shouldn't work.

Reply:
imagine if the testcancel didnt work as expected:

```
                      (1000ms)
[test thread] > . . . . . . . . . . x
                                    ↑
                         exits here ┙
                 (500ms)
[main thread] > . . . . .|- - - - - ?
                         ↑     ↑    ↑
                wakes up ┙     |    |
                   join thread ┙    |
                 test for <= 1000ms ┙
```

if we assume we have a very capable CPU and the instructions + context switches took << 1ms to complete, the time taken will be equal to 1000ms. Under real world scenario the actual time taken will be _very_ close to 1000ms. The current test works, I just thought that the 'safety margin' should be bigger